# Relationships Between Tables

## **Getting Started**

This week we will be learning how to how to pull data from multiple tables.

**Before beginning:** It is recommended that you open/download the _guitar store ERD_ that is available in Github so you can review it during the assignment.

## **Joins**

Joins are how we relate tables during query execution to obtain a specific result. Joins are different from relationships in the sense that relationships exist for the purpose of maintaining data integrity between tables. The relationship that exist in the _guitar store_ database between _Categories_ and _Products_ is in place to ensure that there are no products that don't have a corresponding record in the _Categories_ table. To return related data between these tables we can use the JOIN statement in our SQL.

During this exercise, we will be using the _Products_ and _Categories_ tables. Run the below code cells (2) to see the structure and data in each table.

In [ ]:
--Select all records form gs_products
SELECT *
FROM gs_products

In [ ]:
--Select all records form gs_categories
SELECT *
FROM gs_categories

We can tell from the results of our queries and the _guitar shop_ ERD that instead of storing the category name inside of the product table, we store a category ID and use a relationship/join to look for that category ID in the categories table. There are two main reasons we do this:

- It takes more space to store the word "Basses" than the integer 2. Please see the storage sizes below. This is a massive difference. Instead of storing the word "Basses" everytime that we need it, we can store it one time in the category table and use the ID it was assigned to pull it into our result.
    - 2 in binary: 10
    - "Basses" in binary: 01100010 01100001 01110011 01110011 01100101 01110011
    - So, it takes 24 times as much storage to store the word "Basses" than it does to store the number 2.
- The other reason is that it helps us preserve data integrity. Due to referencial integrity, anything we put in Products.CategoryID must have a matching value in Categories.CategoryID. This prevents erraneous values from sneaking in like "Bassses", "Bases", etc...

### **Inner Join**

Using the JOIN statement in SQL to pull data from multiple tables results in an _inner join._ An inner join will only return records/rows of the selected fields from tables where the joined fields are equal. For example, _Categories_ and _Products_ can be joined on _CategoryID_ since that field is a FK in the _Products_ table. In the below code cell we are selecting all records from both tables where the join fields are equal.

In [ ]:
--Return all records from products where there is a match in categories
SELECT p.*, c.*
FROM gs_products as p 
JOIN gs_categories as c ON p.category_id = c.category_id

When running the above code cell we can see that the result returns all fields from both tables in the order that we entered them into the select statatement. The below code cell contains a query that will just pull in a few fields from each table to evaluate our join.

In [ ]:
--Return specific columns from products where there is a match in categories
SELECT p.product_id, p.product_code, p.category_id AS products_category_id, c.category_id AS categories_category_id, c.category_name
FROM gs_products AS p
JOIN gs_categories AS c ON p.category_id = c.category_id

The above result shows us that we are brining in data from the categories table where Categories.CategoryID matches Products.CategoryID. We know from our earlier example that Category ID 2 is "Basses", so "Basses" is returned as the CategoryName for any products with a 2 in the categoryID. We can also use a WHERE clause to limit the return of data in the products table when joining on the categories table.

In [ ]:
--Return all records from products where the product is a Base
SELECT *
FROM gs_products AS p
JOIN gs_categories AS c ON p.category_id = c.category_id
WHERE c.category_name = 'Basses'

The above code cell returns all products where the category name is "Basses".  Even though we aren't brining in any of the data from the categories table into our select statement, we can still refernce it in our from clause through a join and filter it with criteria in our where clause.

✏️ **Practice**

Edit the below code cell so that the query ruturns all data from the products table that are in the _guitar_ category by using a join.

In [1]:
SELECT *
FROM gs_products AS p
JOIN gs_categories AS c ON p.category_id = c.category_id

(10 rows affected)

Total execution time: 00:00:00.118

product_id category_id product_code product_name product_description list_price discount_percent date_added category_id category_name 1 1 strat Fender Stratocaster The Fender Stratocaster is the electric guitar design that changed the world. New features include a tinted neck, parchment pickguard and control knobs, and a '70s-style logo. Includes select alder body, 21-fret maple neck with your choice of a rosewood or maple fretboard, 3 single-coil pickups, vintage-style tremolo, and die-cast tuning keys. This guitar features a thicker bridge block for increased sustain and a more stable point of contact with the strings. At this low price, why play anything but the real thing?\r\n\r\nFeatures:\r\n\r\n* New features:\r\n* Thicker bridge block\r\n* 3-ply parchment pick guard\r\n* Tinted neck 699.00 30.00 2015-10-30 09:32:40.000 1 Guitars 2 1 les_paul Gibson Les Paul This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, '59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch 1199.00 30.00 2015-12-05 16:33:13.000 1 Guitars 3 1 sg Gibson SG This Gibson SG electric guitar takes the best of the '62 original and adds the longer and sturdier neck joint of the late '60s models. All the classic features you'd expect from a historic guitar. Hot humbuckers go from rich, sweet lightning to warm, tingling waves of sustain. A silky-fast rosewood fretboard plays like a dream. The original-style beveled mahogany body looks like a million bucks. Plus, Tune-O-Matic bridge and chrome hardware. Limited lifetime warranty. Includes hardshell case.\r\n\r\nFeatures:\r\n\r\n* Double-cutaway beveled mahogany body\r\n* Set mahogany neck with rounded '50s profile\r\n* Bound rosewood fingerboard with trapezoid inlays\r\n* Tune-O-Matic bridge with stopbar tailpiece\r\n* Chrome hardware\r\n* 490R humbucker in the neck position\r\n* 498T humbucker in the bridge position\r\n* 2 volume knobs, 2 tone knobs, 3-way switch\r\n* 24-3/4" scale 2517.00 52.00 2016-02-04 11:04:31.000 1 Guitars 4 1 fg700s Yamaha FG700S The Yamaha FG700S solid top acoustic guitar has the ultimate combo for projection and pure tone. The expertly braced spruce top speaks clearly atop the rosewood body. It has a rosewood fingerboard, rosewood bridge, die-cast tuners, body and neck binding, and a tortoise pickguard.\r\n\r\nFeatures:\r\n\r\n* Solid Sitka spruce top\r\n* Rosewood back and sides\r\n* Rosewood fingerboard\r\n* Rosewood bridge\r\n* White/black body and neck binding\r\n* Die-cast tuners\r\n* Tortoise pickguard\r\n* Limited lifetime warranty 489.99 38.00 2016-06-01 11:12:59.000 1 Guitars 5 1 washburn Washburn D10S The Washburn D10S acoustic guitar is superbly crafted with a solid spruce top and mahogany back and sides for exceptional tone. A mahogany neck and rosewood fingerboard make fretwork a breeze, while chrome Grover-style machines keep you perfectly tuned. The Washburn D10S comes with a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n * Spruce top\r\n * Mahogany back, sides\r\n * Mahogany neck Rosewood fingerboard\r\n * Chrome Grover-style machines 299.00 0.00 2016-07-30 13:58:35.000 1 Guitars 6 1 rodriguez Rodriguez Caballero 11 Featuring a carefully chosen, solid Canadian cedar top and laminated bubinga 

**Expected Result**



### **Left/Right Joins**

While inner joins return all records from both tables where there is a match in the joining fields, a left or right join will return all records from the specified table, and only matching records from the joined table. Using the next few code cells, we will build a query that shows all customers who have never placed an order using a left join. Before we do so, run the code cells (2) below to view the data and structure of the customers and orders tables.

In [ ]:
--Return all records from customers
SELECT *
FROM gs_customers

In [ ]:
--Return all records from orders
SELECT *
FROM gs_orders

We can see from the above queries that there are 486 customers and 41 orders. Let's start working on our "Customers without orders" query by creating a query that joins the two tables in customerID to see which customers have placed an order.

In [ ]:
--Return all customers who appear in the orders table
SELECT c.*, o.*
FROM gs_customers as c
JOIN gs_orders as o ON c.customer_id = o.customer_id

The above code cell only returns 41 records, which makes sense since that is how many orders were placed, but we definitely aren't seeing all of our customers since our first set of queries showed 485 different customers. We can use a left join to pull those customers back in. The way that this works is that we are telling the query to return all records from the "left" side of the join statement, so we will leave customers on the left side for now.

In [ ]:
--Use left join to show all customers as well as order information for customers that have ordered
SELECT c.*, o.*
FROM gs_customers as c
LEFT JOIN gs_orders as o ON c.customer_id = o.customer_id

We can see that the query is now showing information for all customers regardless of if they placed an order or not, and there are a few duplicates for customers who placed multiple orders. Please scroll down to customer 37, and then scroll over to view the order data. Notice that all of their order data is Null because they never placed an order. Since there were no orders placed for these customers, there is nothing for the orders table to return here.

What would happen if we made this a right join instead?

In [ ]:
--Use a right join to show all orders as well as customer information if there is a match
SELECT c.*, o.*
FROM gs_customers as c
RIGHT JOIN gs_orders as o ON c.customer_id = o.customer_id

We are back to seeing 41 records and we don't see as many Null values. This is because we are now showing all records on the right side of the join statement (orders) and only those on the left side that have a related record.

Putting together everything that we have learned, we now know enough to create a query that shows which customers haven't placed an order.

- Show all customers- Left join between customers and orders
- Only show customers who don't have an order- set criteria to Null for order data

Let's incoroporate this into our next query.

In [ ]:
--Use left join to show only customers who have never placed an order
SELECT c.*, o.*
FROM gs_customers as c
LEFT JOIN gs_orders as o ON c.customer_id = o.customer_id
WHERE o.customer_id is NULL

By adding a where clause to only show Nulls from Orders.CustomerID we were able to filter out any customers that had related data in the orders table, or in other words, had placed an order. Now we know that there are 450 customers in our database that have yet to place an order.

✏️ **Practice**

Edit the below query to show all customers who do not have an address in the addresses table.

In [5]:
SELECT c.*
FROM gs_customers AS c
JOIN gs_addresses AS a ON c.customer_id = a.customer_id
WHERE a.line1 is NULL

(512 rows affected)

Total execution time: 00:00:00.057

address_id customer_id line1 line2 city state_code zip_code phone disabled_address 1 1 100 East Ridgewood Ave. Paramus NJ 07652 201-653-4472 0 2 1 21 Rosewood Rd. Woodcliff Lake NJ 07677 201-653-4472 0 3 2 16285 Wendell St. Omaha NE 68135 402-896-2576 0 4 3 19270 NW Cornell Rd. Beaverton OR 97006 503-654-1291 0 5 4 186 Vermont St. Apt. 2 San Francisco CA 94110 415-292-6651 0 6 4 1374 46th Ave. San Francisco CA 94129 415-292-6651 0 7 5 6982 Palm Ave. Fresno CA 93711 559-431-2398 0 8 6 23 Mountain View St. Denver CO 80208 303-912-3852 0 9 7 7361 N. 41st St. Apt. B New York NY 10012 212-335-2093 0 10 7 3829 Broadway Ave. Suite 2 New York NY 10012 212-239-1208 0 11 8 2381 Buena Vista St. Los Angeles CA 90023 213-772-5033 0 12 8 291 W. Hollywood Blvd. Los Angeles CA 90024 213-391-2938 0 13 9 96950 Hidden Ln Aberdeen MD 21001 410-259-2118 0 14 10 6980 Dorsett Rd Abilene KS 67410 785-219-7724 0 15 11 33 State St Abilene TX 79601 325-667-7868 0 16 11 5 E Truman Rd Abilene TX 79602 325-740-3778 0 17 12 8284 Hart St Abilene KS 67410 785-253-7049 0 18 13 9387 Charcot Ave Absecon NJ 8201 609-234-8376 0 19 14 3777 E Richmond St #900 Akron OH 44302 330-700-2312 0 20 15 74 S Westgate St Albany NY 12204 518-448-8982 0 21 15 9390 S Howell Ave Albany GA 31701 229-365-9658 0 22 16 89992 E 15th St Alliance NE 69301 308-250-6987 0 23 17 812 S Haven St Amarillo TX 79109 806-558-5848 0 24 18 72119 S Walker Ave #63 Anaheim CA 92801 714-663-9740 0 25 19 639 Main St Anchorage AK 99501 907-921-2010 0 26 19 18 Fountain St Anchorage AK 99515 907-873-2882 0 27 20 1747 Calle Amanecer #2 Anchorage AK 99501 907-914-9482 0 28 21 735 Crawford Dr Anchorage AK 99501 907-770-3542 0 29 22 560 Civic Center Dr Ann Arbor MI 48103 734-408-8174 0 30 23 747 Leonis Blvd Annandale VA 22003 703-874-4248 0 31 24 70099 E North Ave Arlington TX 76013 817-947-9480 0 32 25 73 W Barstow Ave Arlington Heights IL 60004 847-613-5866 0 33 26 3 Mcauley Dr Ashland OH 44805 419-800-6759 0 34 27 6 S 33rd St Aston PA 19014 610-492-4643 0 35 28 92 Broadway Astoria NY 11103 718-728-5051 0 36 29 45 2nd Ave #9759 Atlanta GA 30328 770-531-2842 0 37 29 26 Montgomery St Atlanta GA 30328 770-930-9967 0 38 30 187 Market St Atlanta GA 30342 404-607-8435 0 39 31 82 Winsor St #54 Atlanta GA 30340 770-584-4119 0 40 32 92 Main St Atlantic City NJ 8401 609-263-9243 0 41 32 506 S Hacienda Dr Atlantic City NJ 8401 609-854-7156 0 42 33 59 N Groesbeck Hwy Austin TX 78731 512-861-3814 0 43 33 2026 N Plankinton Ave #3 Austin TX 78754 512-693-8345 0 44 34 3125 Packer Ave #9851 Austin TX 78753 512-742-1149 0 45 35 228 Runamuck Pl #2808 Baltimore MD 21224 410-804-4694 0 46 36 6 Kains Ave Baltimore MD 21215 410-957-6903 0 47 37 2 S Biscayne Blvd Baltimore MD 21230 410-773-3862 0 48 38 4 Iwaena St Baltimore MD 21202 410-429-4888 0 49 39 2 W Scyene Rd #3 Baltimore MD 21217 410-522-7621 0 50 40 37855 Nolan Rd Bangor ME 4401 207-233-6185 0 51 40 34 Saint George Ave #2 Bangor ME 4401 207-748-3722 0 52 41 95 Main Ave #2 Barberton OH 44203 330-566-8898 0 53 42 70 W Main St Beachwood OH 44122 216-733-8494 0 54 43 3 Elmwood Dr Beaverton OR 97005 503-707-5812 0 55 44 34 Raritan Center Pky Bellflower CA 90706 562-719-7922 0 56 45 64 Lakeview Ave Beloit WI 53511 608-942-8836 0 57 46 2887 Knowlton St #5435 Berkeley CA 94710 510-942-5916 0 58 47 429 Tiger Ln Beverly Hills CA 90212 310-969-7230 0 59 47 598 43rd St Beverly Hills CA 90210 310-652-2379 0 60 48 5 Elmwood Park Blvd Biloxi MS 39530 228-432-4635 0 61 49 9506 Edgemore Ave Bladensburg MD 20710 301-591-3034 0 62 50 539 Coldwater Canyon Ave Bloomfield NJ 7003 973-473-5108 0 63 51 772 W River Dr Bloomington IN 47404 812-442-8544 0 64 52 70 Euclid Ave #722 Bohemia NY 11716 631-291-4976 0 65 53 74 W College St Boise ID 83707 208-737-8439 0 66 53 9635 S Main St Boise ID 83704 208-690-3315 0 67 54 67 Rv Cent Boise ID 83709 208-206-9848 0 68 55 38938 Park Blvd Boston MA 2128 617-997-5771 0 69 55 8 County Center Dr #647 Boston MA 2210 617-697-6024 0 70 56 5 S Colorado Blvd #449 

## **Junction/Join Tables**

Junciton, or join, tables are a special type of table that helps us resolve a many-to-many relationship. In the _guitar store_ example we have a many-to-many relationship between products and orders. This is because an individual product can be part of many orders, and an order can be made of many products. For example, Order #31 may contain products 1 and 6. At first this looks like a one-to-many relationship, however, Order # 41 contains products 6 and 8. We have multiple products appearing in different orders, which makes this a many-to-many relationship. To resolve this, we use a junction table. This junctino table, _OrderItems_, lists the OrderID and each associated ProductID in their own row so that we can query them. If we run the below code cells (2) we can see that there are no fields that we can use to join Prducts and Orders. They have no common data.

In [ ]:
SELECT *
FROM gs_products AS p

In [ ]:
SELECT *
FROM gs_orders as o

Now, if we look at the _OrderItems_ table then we can see that it contains information from both tables. This is showing which items (item\_id) were part of which orders (order\_id)

In [ ]:
SELECT *
FROM gs_order_items AS oi

This table also includes useful information such as what the discount on the product was at the time of the order, the price that was paid, and the quantity purchased. We can think of this as a transaction table that is storing transaction level data for each product in each order. It is the most granular stage of our data.

If we want to connect products to orders, then we will need to join each of our products and orders tables to the OrderItems table. This can be done by using multiple join statements like in the cell below. At the end of one join statement, we just start the next one. The below code cell is showing the productID and orderID for each product/order pairing. If you scroll up, you will notice that these are the same combinations as what is in the ItemOrders table, but we are pulling from the Products and Orders table now instead!

In [ ]:
SELECT p,product_id, o.order_id
FROM gs_products as p
JOIN gs_order_items as oi ON p.product_id = oi.product_id
JOIN gs_orders as o ON oi.order_id = o.order_id

**✏️ Practice**

Edit the below code cell so that you are returning the name of each product, as well as the customerID of the customer who placed the order, and the OrderDate.

In [7]:
SELECT p.product_name, o.customer_id, o.order_date
FROM gs_products as p
JOIN gs_order_items as oi ON p.product_id = oi.product_id
JOIN gs_orders as o ON oi.order_id = o.order_id

(47 rows affected)

Total execution time: 00:00:00.050

product_name customer_id order_date Gibson Les Paul 1 2016-03-28 09:40:28.000 Hofner Icon 2 2016-03-28 11:23:20.000 Fender Stratocaster 1 2016-03-29 09:44:58.000 Ludwig 5-piece Drum Set with Cymbals 1 2016-03-29 09:44:58.000 Gibson Les Paul 3 2016-03-30 15:22:31.000 Tama 5-Piece Drum Set with Cymbals 4 2016-03-31 05:43:11.000 Tama 5-Piece Drum Set with Cymbals 5 2016-03-31 18:37:22.000 Washburn D10S 6 2016-04-01 23:11:12.000 Gibson SG 6 2016-04-01 23:11:12.000 Washburn D10S 6 2016-04-01 23:11:12.000 Yamaha FG700S 7 2016-04-02 11:26:38.000 Rodriguez Caballero 11 4 2016-04-03 12:22:31.000 Fender Precision 8 2016-04-03 14:59:20.000 Rodriguez Caballero 11 9 2016-04-04 06:24:44.000 Fender Stratocaster 10 2016-04-04 08:15:12.000 Yamaha FG700S 11 2016-04-04 11:20:31.000 Gibson Les Paul 12 2016-04-05 09:24:53.000 Gibson SG 13 2016-04-05 14:52:17.000 Rodriguez Caballero 11 14 2016-04-06 07:53:42.000 Ludwig 5-piece Drum Set with Cymbals 14 2016-04-06 07:53:42.000 Rodriguez Caballero 11 15 2016-04-06 17:24:28.000 Rodriguez Caballero 11 16 2016-04-06 18:41:53.000 Gibson Les Paul 17 2016-04-08 12:21:31.000 Yamaha FG700S 18 2016-04-10 09:33:23.000 Gibson SG 19 2016-04-11 08:21:32.000 Fender Stratocaster 8 2016-04-12 12:26:52.000 Rodriguez Caballero 11 20 2016-04-14 07:59:31.000 Rodriguez Caballero 11 21 2016-04-17 17:40:22.000 Gibson Les Paul 22 2016-04-20 08:23:32.000 Rodriguez Caballero 11 23 2016-04-20 08:14:45.000 Fender Stratocaster 24 2016-04-20 09:17:52.000 Washburn D10S 25 2016-04-21 17:52:24.000 Gibson SG 4 2016-04-25 23:36:41.000 Fender Precision 26 2016-04-27 16:21:31.000 Fender Stratocaster 27 2016-04-29 06:47:14.000 Rodriguez Caballero 11 27 2016-04-29 06:47:14.000 Fender Stratocaster 18 2016-05-01 01:23:23.000 Tama 5-Piece Drum Set with Cymbals 28 2016-05-01 09:11:51.000 Gibson Les Paul 29 2016-05-02 11:36:12.000 Gibson SG 30 2016-05-04 03:52:23.000 Hofner Icon 31 2016-05-04 12:31:33.000 Fender Stratocaster 32 2016-05-06 14:15:21.000 Gibson Les Paul 33 2016-05-08 11:41:24.000 Rodriguez Caballero 11 29 2016-05-08 22:22:26.000 Rodriguez Caballero 11 34 2016-05-08 21:41:29.000 Hofner Icon 35 2016-05-09 07:52:55.000 Rodriguez Caballero 11 35 2016-05-09 07:52:55.000

## **Exercises**

If you have not already, I would strongly recommend having access to the ERD Linked in Canvas for the Exercises.

1\. Using the below code cell, return the order date, ship amount, and ship date for every order, as well as the name of the customer that placed the order.

**First 5 rows of correct solution**
|order_date|ship_amount|ship_date|first_name|last_name|
|---|---|---|---|---|
|2016-03-28 09:40:28.000|5.00|2016-03-31 09:41:11.000|Allan|Sherwood|
|2016-03-28 11:23:20.000|5.00|2016-03-31 11:24:03.000|Barry|Zimmer|
|2016-03-29 09:44:58.000|10.00|2016-04-01 09:45:41.000|Allan|Sherwood|
|2016-03-30 15:22:31.000|10.00|2016-04-02 15:23:14.000|Christine|Brown|
|2016-03-31 05:43:11.000|5.00|2016-04-03 05:43:54.000|David|Goldstein|

In [8]:
--Insert your code below this line. You can make your own comments by using two hyphens
SELECT o.order_date, o.ship_amount, o.ship_date, c.first_name, c.last_name
FROM gs_orders AS o
JOIN gs_customers AS c ON o.customer_id = c.customer_id


(41 rows affected)

Total execution time: 00:00:00.049

order_date ship_amount ship_date first_name last_name 2016-03-28 09:40:28.000 5.00 2016-03-31 09:41:11.000 Allan Sherwood 2016-03-28 11:23:20.000 5.00 2016-03-31 11:24:03.000 Barry Zimmer 2016-03-29 09:44:58.000 10.00 2016-04-01 09:45:41.000 Allan Sherwood 2016-03-30 15:22:31.000 10.00 2016-04-02 15:23:14.000 Christine Brown 2016-03-31 05:43:11.000 5.00 2016-04-03 05:43:54.000 David Goldstein 2016-03-31 18:37:22.000 5.00 2016-04-03 18:38:05.000 Erin Valentino 2016-04-01 23:11:12.000 15.00 2016-04-04 23:11:55.000 Frank Lee Wilson 2016-04-02 11:26:38.000 5.00 2016-04-05 11:27:21.000 Gary Hernandez 2016-04-03 12:22:31.000 15.00 2016-04-06 12:23:14.000 David Goldstein 2016-04-03 14:59:20.000 5.00 2016-04-06 15:00:03.000 Heather Esway 2016-04-04 06:24:44.000 5.00 2016-04-07 06:25:27.000 James Butt 2016-04-04 08:15:12.000 5.00 2016-04-07 08:15:55.000 Josephine Darakjy 2016-04-04 11:20:31.000 5.00 2016-04-07 11:21:14.000 Art Venere 2016-04-05 09:24:53.000 10.00 2016-04-08 09:25:36.000 Lenna Paprocki 2016-04-05 14:52:17.000 5.00 2016-04-08 14:53:00.000 Donette Foller 2016-04-06 07:53:42.000 10.00 2016-04-09 07:54:25.000 Simona Morasca 2016-04-06 17:24:28.000 5.00 2016-04-09 17:25:11.000 Mitsue Tollner 2016-04-06 18:41:53.000 5.00 2016-04-09 18:42:36.000 Leota Dilliard 2016-04-08 12:21:31.000 10.00 2016-04-11 12:22:14.000 Sage Wieser 2016-04-10 09:33:23.000 5.00 2016-04-13 09:34:06.000 Kris Marrier 2016-04-11 08:21:32.000 10.00 2016-04-14 08:22:15.000 Minna Amigon 2016-04-12 12:26:52.000 5.00 2016-04-15 12:27:35.000 Heather Esway 2016-04-14 07:59:31.000 5.00 2016-04-17 08:00:14.000 Abel Maclead 2016-04-17 17:40:22.000 5.00 2016-04-20 17:41:05.000 Kiley Caldarera 2016-04-20 08:23:32.000 10.00 2016-04-23 08:24:15.000 Graciela Ruta 2016-04-20 08:14:45.000 5.00 2016-04-23 08:15:28.000 Cammy Albares 2016-04-20 09:17:52.000 5.00 2016-04-23 09:18:35.000 Mattie Poquette 2016-04-21 17:52:24.000 5.00 2016-04-24 17:53:07.000 Meaghan Garufi 2016-04-25 23:36:41.000 25.00 2016-04-28 23:37:24.000 David Goldstein 2016-04-27 16:21:31.000 5.00 2016-04-30 16:22:14.000 Gladys Rim 2016-04-29 06:47:14.000 10.00 2016-05-02 06:47:57.000 Yuki Whobrey 2016-05-01 01:23:23.000 5.00 NULL Kris Marrier 2016-05-01 09:11:51.000 10.00 2016-05-04 09:12:34.000 Fletcher Flosi 2016-05-02 11:36:12.000 5.00 2016-05-05 11:36:55.000 Bette Nicka 2016-05-04 03:52:23.000 5.00 2016-05-07 03:53:06.000 Veronika Inouye 2016-05-04 12:31:33.000 5.00 2016-05-07 12:32:16.000 Willard Kolmetz 2016-05-06 14:15:21.000 5.00 2016-05-09 14:16:04.000 Maryann Royster 2016-05-08 11:41:24.000 10.00 NULL Alisha Slusarski 2016-05-08 22:22:26.000 5.00 NULL Bette Nicka 2016-05-08 21:41:29.000 5.00 NULL Allene Iturbide 2016-05-09 07:52:55.000 10.00 NULL Chanel Caudy

2\. Using the below code cell, return the first name, last name, and email address for each customer who has placed and order and has a yahoo email address.

**First 5 rows of correct solution**
|first_name|last_name|email_address|
|---|---|---|
|Allan|Sherwood|allan.sherwood@yahoo.com|
|Allan|Sherwood|allan.sherwood@yahoo.com|
|Gary|Hernandez|gary_hernandez@yahoo.com|
|Mitsue|Tollner|mitsue_tollner@yahoo.com|
|Minna|Amigon|minna_amigon@yahoo.com|

In [9]:
--Insert your code below this line. You can make your own comments by using two hyphens
SELECT c.first_name, c.last_name, c.email_address
FROM gs_customers AS c
WHERE c.email_address LIKE '%@yahoo.com'

(85 rows affected)

Total execution time: 00:00:00.050

first_name last_name email_address Allan Sherwood allan.sherwood@yahoo.com Gary Hernandez gary_hernandez@yahoo.com Mitsue Tollner mitsue_tollner@yahoo.com Minna Amigon minna_amigon@yahoo.com Fletcher Flosi fletcher.flosi@yahoo.com Willow Kusko wkusko@yahoo.com Jose Stockham jose@yahoo.com Kanisha Waycott kanisha_waycott@yahoo.com Blair Malet bmalet@yahoo.com Brock Bolognia bbolognia@yahoo.com Karl Klonowski karl_klonowski@yahoo.com Moon Parlato moon@yahoo.com Elza Lipke elza@yahoo.com Lettie Isenhower lettie_isenhower@yahoo.com Malinda Hochard malinda.hochard@yahoo.com Scarlet Cartan scarlet.cartan@yahoo.com Oretha Menter oretha_menter@yahoo.com Thurman Manno thurman.manno@yahoo.com Beatriz Corrington beatriz@yahoo.com Marti Maybury marti.maybury@yahoo.com Carissa Batman carissa.batman@yahoo.com Gregoria Pawlowicz gpawlowicz@yahoo.com Chantell Maynerich chantell@yahoo.com Dierdre Yum dyum@yahoo.com Dorthy Hidvegi dhidvegi@yahoo.com Fannie Lungren fannie.lungren@yahoo.com Jennifer Fallick jfallick@yahoo.com Eun Coody eun@yahoo.com Helga Fredicks helga_fredicks@yahoo.com Fausto Agramonte fausto_agramonte@yahoo.com Ressie Auffrey ressie.auffrey@yahoo.com Justine Mugnolo jmugnolo@yahoo.com Marvel Raymo mraymo@yahoo.com Deandrea Hughey deandrea@yahoo.com Kimberlie Duenas kimberlie_duenas@yahoo.com Skye Fillingim skye_fillingim@yahoo.com Jade Farrar jade.farrar@yahoo.com Stevie Westerbeck stevie.westerbeck@yahoo.com Buddy Cloney buddy.cloney@yahoo.com Matthew Neither mneither@yahoo.com Raul Upthegrove rupthegrove@yahoo.com Kayleigh Lace kayleigh.lace@yahoo.com Christiane Eschberger christiane.eschberger@yahoo.com Goldie Schirpke goldie.schirpke@yahoo.com Amie Perigo amie.perigo@yahoo.com Helaine Halter hhalter@yahoo.com France Buzick france.buzick@yahoo.com Jerry Dallen jerry.dallen@yahoo.com Leonora Mauson leonora@yahoo.com Dulce Labreche dulce_labreche@yahoo.com Kate Keneipp kate_keneipp@yahoo.com Nobuko Halsey nobuko.halsey@yahoo.com Margart Meisel margart_meisel@yahoo.com Kristofer Bennick kristofer.bennick@yahoo.com Elouise Gwalthney egwalthney@yahoo.com Maile Linahan mlinahan@yahoo.com Markus Lukasik markus@yahoo.com Ciara Ventura cventura@yahoo.com Galen Cantres galen@yahoo.com Truman Feichtner tfeichtner@yahoo.com Lizbeth Kohl lizbeth@yahoo.com Lashandra Klang lashandra@yahoo.com Laurel Pagliuca laurel@yahoo.com Annelle Tagala annelle@yahoo.com Mollie Mcdoniel mollie_mcdoniel@yahoo.com Tasia Andreason tasia_andreason@yahoo.com Arthur Farrow arthur.farrow@yahoo.com Mitzie Hudnall mitzie_hudnall@yahoo.com Bernardine Rodefer bernardine_rodefer@yahoo.com Janine Rhoden jrhoden@yahoo.com Eden Jayson eden_jayson@yahoo.com Golda Kaniecki golda_kaniecki@yahoo.com Jolene Ostolaza jolene@yahoo.com Keneth Borgman keneth@yahoo.com Elke Sengbusch elke_sengbusch@yahoo.com Trinidad Mcrae trinidad_mcrae@yahoo.com Mari Lueckenbach mari_lueckenbach@yahoo.com Dick Wenzinger dick@yahoo.com Iluminada Ohms iluminada.ohms@yahoo.com Quentin Swayze quentin_swayze@yahoo.com Jeanice Claucherty jeanice.claucherty@yahoo.com Jacqueline Rowling jacqueline.rowling@yahoo.com Colette Kardas colette.kardas@yahoo.com Sharen Bourbon sbourbon@yahoo.com Glory Schieler glory@yahoo.com

3\. Using the below code cell, return the order\_id and order date for items ordered by Kris Marrier

**First 5 rows of correct solution**

| order\_id | order\_date |
| --- | --- |
| 20 | 2016-04-10 09:33:23.000 |
| 32 | 2016-05-01 01:23:23.000 |

In [10]:
--Insert your code below this line. You can make your own comments by using two hyphens
SELECT o.order_id, o.order_date
FROM gs_orders AS o
JOIN gs_customers AS c ON o.customer_id = c.customer_id
WHERE c.first_name = 'Kris' AND c.last_name = 'Marrier'

(2 rows affected)

Total execution time: 00:00:00.049

order_id order_date 20 2016-04-10 09:33:23.000 32 2016-05-01 01:23:23.000

4\. Expand the query from exercise 3 to also include the product name, price, and quantity. You will have to join a total of four tables.

**First 5 rows of correct solution**
|order_id|order_date|product_name|item_price|quantity|
|---|---|---|---|---|
|20|2016-04-10 09:33:23.000|Yamaha FG700S|799.99|1|
|32|2016-05-01 01:23:23.000|Fender Stratocaster|2517.00|1|


In [12]:
--Insert your code below this line. You can make your own comments by using two hyphens
SELECT o.order_id, o.order_date, p.product_name, oi.item_price, oi.quantity
FROM gs_orders AS o
JOIN gs_customers AS c ON o.customer_id = c.customer_id
JOIN gs_order_items AS oi ON o.order_id = oi.order_id
JOIN gs_products AS p ON oi.product_id = p.product_id
WHERE c.first_name = 'Kris' AND c.last_name = 'Marrier'


(2 rows affected)

Total execution time: 00:00:00.060

order_id order_date product_name item_price quantity 20 2016-04-10 09:33:23.000 Yamaha FG700S 799.99 1 32 2016-05-01 01:23:23.000 Fender Stratocaster 2517.00 1

## **Scenario**

The owner of Guitar Store wants you to create a query that shows all of the needed information for creating a batch of invoices. Your query should contain the Cateegory Name, Products Name, Description, List Price, Item Price, Discount Amount, Quantity, OrderDate, Shipping and tax costs, Customer name, Customer Shipping address, and their email.

Create your query in the code cell below:

**First 5 rows of correct solution**
|category_name|product_name|product_description|list_price|item_price|discount_amount|quantity|order_date|ship_amount|tax_amount|first_name|last_name|Line1|line2|city|state_code|zip_code|email_address|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
|Guitars|Gibson Les Paul|This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, &#39;59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch|1199.00|1199.00|359.70|1|2016-03-28 09:40:28.000|5.00|58.75|Allan|Sherwood|100 East Ridgewood Ave.||Paramus|NJ|07652|allan.sherwood@yahoo.com|
|Basses|Hofner Icon|With authentic details inspired by the original, the Hofner Icon makes the legendary violin bass available to the rest of us. Don&#39;t get the idea that this a just a &quot;nowhere man&quot; look-alike. This quality instrument features a real spruce top and beautiful flamed maple back and sides. The semi-hollow body and set neck will give you the warm, round tone you expect from the violin bass.\r\n\r\nFeatures:\r\n\r\n* Authentic details inspired by the original\r\n* Spruce top\r\n* Flamed maple back and sides\r\n* Set neck\r\n* Rosewood fretboard\r\n* 30&quot; scale\r\n* 22 frets\r\n* Dot inlay|499.99|489.99|186.20|1|2016-03-28 11:23:20.000|5.00|21.27|Barry|Zimmer|16285 Wendell St.||Omaha|NE|68135|barryz@gmail.com|
|Guitars|Fender Stratocaster|The Fender Stratocaster is the electric guitar design that changed the world. New features include a tinted neck, parchment pickguard and control knobs, and a &#39;70s-style logo. Includes select alder body, 21-fret maple neck with your choice of a rosewood or maple fretboard, 3 single-coil pickups, vintage-style tremolo, and die-cast tuning keys. This guitar features a thicker bridge block for increased sustain and a more stable point of contact with the strings. At this low price, why play anything but the real thing?\r\n\r\nFeatures:\r\n\r\n* New features:\r\n* Thicker bridge block\r\n* 3-ply parchment pick guard\r\n* Tinted neck|699.00|2517.00|1308.84|1|2016-03-29 09:44:58.000|10.00|102.29|Allan|Sherwood|100 East Ridgewood Ave.||Paramus|NJ|07652|allan.sherwood@yahoo.com|
|Drums|Ludwig 5-piece Drum Set with Cymbals|This product includes a Ludwig 5-piece drum set and a Zildjian starter cymbal pack.\r\n\r\nWith the Ludwig drum set, you get famous Ludwig quality. This set features a bass drum, two toms, a floor tom, and a snare?each with a wrapped finish. Drum hardware includes LA214FP bass pedal, snare stand, cymbal stand, hi-hat stand, and a throne.\r\n\r\nWith the Zildjian cymbal pack, you get a 14&quot; crash, 18&quot; crash/ride, and a pair of 13&quot; hi-hats. Sound grooves and round hammer strikes in a simple circular pattern on the top surface of these cymbals magnify the basic sound of the distinctive alloy.\r\n\r\nFeatures:\r\n\r\n* Famous Ludwig quality\r\n* Wrapped finishes\r\n* 22&quot; x 16&quot; kick drum\r\n* 12&quot; x 10&quot; and 13&quot; x 11&quot; toms\r\n* 16&quot; x 16&quot; floor tom\r\n* 14&quot; x 6-1/2&quot; snare drum kick pedal\r\n* Snare stand\r\n* Straight cymbal stand hi-hat stand\r\n* FREE throne|699.99|415.00|161.85|1|2016-03-29 09:44:58.000|10.00|102.29|Allan|Sherwood|100 East Ridgewood Ave.||Paramus|NJ|07652|allan.sherwood@yahoo.com|
|Guitars|Gibson Les Paul|This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, &#39;59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch|1199.00|1199.00|359.70|2|2016-03-30 15:22:31.000|10.00|117.50|Christine|Brown|19270 NW Cornell Rd.||Beaverton|OR|97006|christineb@solarone.com|


In [17]:
--Insert your code below this line. You can make your own comments by using two hyphens
SELECT cat.category_name, p.product_name, p.product_description, p.list_price, oi.item_price, oi.discount_amount, oi.quantity, o.ship_amount, o.tax_amount, c.first_name, c.last_name, a.line1, a.line2, a.city, a.state_code, a.zip_code, c.email_address
FROM gs_categories AS cat
JOIN gs_products AS p ON cat.category_id = p.category_id
JOIN gs_order_items AS oi ON p.product_id = oi.product_id
JOIN gs_orders AS o ON oi.order_id = o.order_id
JOIN gs_customers AS c ON o.customer_id = c.customer_id
JOIN gs_addresses AS a ON c.customer_id = a.customer_id

(63 rows affected)

Total execution time: 00:00:00.060

category_name product_name product_description list_price item_price discount_amount quantity ship_amount tax_amount first_name last_name line1 line2 city state_code zip_code email_address Guitars Gibson Les Paul This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, '59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch 1199.00 1199.00 359.70 1 5.00 58.75 Allan Sherwood 100 East Ridgewood Ave. Paramus NJ 07652 allan.sherwood@yahoo.com Guitars Fender Stratocaster The Fender Stratocaster is the electric guitar design that changed the world. New features include a tinted neck, parchment pickguard and control knobs, and a '70s-style logo. Includes select alder body, 21-fret maple neck with your choice of a rosewood or maple fretboard, 3 single-coil pickups, vintage-style tremolo, and die-cast tuning keys. This guitar features a thicker bridge block for increased sustain and a more stable point of contact with the strings. At this low price, why play anything but the real thing?\r\n\r\nFeatures:\r\n\r\n* New features:\r\n* Thicker bridge block\r\n* 3-ply parchment pick guard\r\n* Tinted neck 699.00 2517.00 1308.84 1 10.00 102.29 Allan Sherwood 100 East Ridgewood Ave. Paramus NJ 07652 allan.sherwood@yahoo.com Drums Ludwig 5-piece Drum Set with Cymbals This product includes a Ludwig 5-piece drum set and a Zildjian starter cymbal pack.\r\n\r\nWith the Ludwig drum set, you get famous Ludwig quality. This set features a bass drum, two toms, a floor tom, and a snare?each with a wrapped finish. Drum hardware includes LA214FP bass pedal, snare stand, cymbal stand, hi-hat stand, and a throne.\r\n\r\nWith the Zildjian cymbal pack, you get a 14" crash, 18" crash/ride, and a pair of 13" hi-hats. Sound grooves and round hammer strikes in a simple circular pattern on the top surface of these cymbals magnify the basic sound of the distinctive alloy.\r\n\r\nFeatures:\r\n\r\n* Famous Ludwig quality\r\n* Wrapped finishes\r\n* 22" x 16" kick drum\r\n* 12" x 10" and 13" x 11" toms\r\n* 16" x 16" floor tom\r\n* 14" x 6-1/2" snare drum kick pedal\r\n* Snare stand\r\n* Straight cymbal stand hi-hat stand\r\n* FREE throne 699.99 415.00 161.85 1 10.00 102.29 Allan Sherwood 100 East Ridgewood Ave. Paramus NJ 07652 allan.sherwood@yahoo.com Guitars Gibson Les Paul This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, '59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch 1199.00 1199.00 359.70 1 5.00 58.75 Allan Sherwood 2